internal/operator-controller/rukpak/bundle/config.go

```diff
@@ -6,119 +6,489 @@ import (
+func buildBundleConfigSchema(installModes sets.Set[v1alpha1.InstallMode]) (string, error) {
+	schema := map[string]interface{}{
+		"$schema":              "http://json-schema.org/draft-07/schema#",
+		"type":                 "object",
+		"additionalProperties": false, // Reject unknown fields (catches typos and misconfigurations)
+	}
+
+	properties := map[string]interface{}{}
+	var requiredFields []string
+
+	// Add watchNamespace property if the bundle supports it
+	if isWatchNamespaceConfigurable(installModes) {
+		watchNSProperty := buildWatchNamespaceProperty(installModes, &requiredFields)
+		properties["watchNamespace"] = watchNSProperty
+	}
+
+	schema["properties"] = properties
+	if len(requiredFields) > 0 {
+		schema["required"] = requiredFields
 	}
 
-	// if the bundle does not support the watchNamespace configuration and it is set, treat it like any unknown field
-	if config.WatchNamespace != nil && !isWatchNamespaceConfigSupported(bundleInstallModeSet) {
-		return errors.New(`unknown field "watchNamespace"`)
+	schemaJSON, err := json.Marshal(schema)
+	if err != nil {
+		return "", fmt.Errorf("failed to serialize schema: %w", err)
 	}
 
-	// if watchNamespace is required then ensure that it is set
-	if config.WatchNamespace == nil && isWatchNamespaceConfigRequired(bundleInstallModeSet) {
-		return errors.New(`required field "watchNamespace" is missing`)
+	return string(schemaJSON), nil
+}
+
+// buildWatchNamespaceProperty builds the watchNamespace schema property.
+//
+//   - When bundle supports AllNamespaces -> watchNamespace optional (can be null)
+//   - When bundle only supports Single/Own -> watchNamespace required
+//   - When OwnNamespace only -> must equal install namespace
+//   - When SingleNamespace only -> must NOT equal install namespace
+func buildWatchNamespaceProperty(installModes sets.Set[v1alpha1.InstallMode], requiredFields *[]string) map[string]interface{} {
+	watchNSProperty := map[string]interface{}{
+		"description": "The namespace that the operator should watch for custom resources",
 	}
 
-	// if watchNamespace is set then ensure it is a valid namespace
-	if config.WatchNamespace != nil {
-		if errs := validation.IsDNS1123Subdomain(*config.WatchNamespace); len(errs) > 0 {
-			return fmt.Errorf("invalid 'watchNamespace' %q: namespace must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character", *config.WatchNamespace)
+	hasOwnNamespace := installModes.Has(v1alpha1.InstallMode{Type: v1alpha1.InstallModeTypeOwnNamespace, Supported: true})
+	hasSingleNamespace := installModes.Has(v1alpha1.InstallMode{Type: v1alpha1.InstallModeTypeSingleNamespace, Supported: true})
+
+	format := selectNamespaceFormat(hasOwnNamespace, hasSingleNamespace)
+
+	if isWatchNamespaceConfigRequired(installModes) {
+		watchNSProperty["type"] = "string"
+		watchNSProperty["format"] = format
+		*requiredFields = append(*requiredFields, "watchNamespace")
```

Comment:
it looks like that we have a side-effect in the function by appending fields to passed `requiredFields` arg, correct?